mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp

```diff
@@ -136,6 +136,38 @@ struct FoldEmptyTensorWithUnPackOp : public OpRewritePattern<UnPackOp> {
   }
 };
 
+// Fold concat operation where all the operands are empty.
+struct FoldConcatsOfEmpty : public OpRewritePattern<ConcatOp> {
+  using OpRewritePattern<ConcatOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ConcatOp concatOp,
+                                PatternRewriter &rewriter) const override {
+    auto concatOperands = concatOp.getInputs();
+    if (concatOperands.empty()) {
+      return failure();
+    }
+    auto firstEmptyOp = concatOperands.front().getDefiningOp<tensor::EmptyOp>();
+    if (!firstEmptyOp) {
+      return failure();
+    }
+    auto isDefinedByEmptyOp = [](Value v) -> bool {
+      return v.getDefiningOp<tensor::EmptyOp>();
+    };
+    if (!llvm::all_of(concatOperands.drop_front(), isDefinedByEmptyOp)) {
+      return rewriter.notifyMatchFailure(
+          concatOp, "not all operands are defined by an empty op");
+    }
```

Comment:
Why is the first operand special? Can't we just write something like below?

```cpp
if (!llvm::all_of(concatOperands, isDefinedByEmptyOp)) {
  ....
}
```

Reply:
Just adding an early exit if the first operand is itself is not an empty op.

Reply:
I thought that the `llvm::all_of` should already consider the case? I imagine that it also does early-exit if the first operand is not an empty op. It could be different in multi-threads though, idk.

Reply:
+1 I don't quite understand the code here: llvm::all_of should already be having the early exit.

Reply:
I was just trying to avoid even doing the `all_of` at all if even the first operand is not empty. But ok, I'll change it.